apps/settings/js/modules/bluetooth/bluetooth_connection_manager.js

```diff
@@ -0,0 +1,868 @@
+              JSON.stringify(collectedConnectedDevicesByProfile));
+        return Promise.resolve(collectedConnectedDevicesByProfile);
+      });
+    },
+
+    /**
+     * The method will get connected device by inputed profile.
+     *
+     * @access private
+     * @memberOf BluetoothConnectionManager
+     * @param {String} profileID
+     * @returns {Promise} resolve: the connected devices in array
+     * @returns {Promise} reject: reason
+     */
+    _getConnectedDevicesByProfile: 
+    function btcm__getConnectedDevicesByProfile(profileID) {
+      return new Promise((resolve, reject) => {
+        if (!this._defaultAdapter) {
+          reject('default adapter is not existed!!');
+        }
+
+        var req = this._defaultAdapter.getConnectedDevices(profileID);
+        req.onsuccess = () => {
+          Debug('_getConnectedDevicesByProfile(): onsuccess(): req.result = ' + 
+                JSON.stringify(req.result));
+          resolve(req.result || []);
+        };
+
+        req.onerror = () => {
+          var reason = 'getConnectedDevices(): onerror';
+          Debug('_getConnectedDevicesByProfile(): onerror(): Reject:' + reason);
+          reject(reason);
+        };
+      });
+    },
+
+    /**
+     * Get device from paired devices by address.
+     *
+     * @access private
+     * @memberOf BluetoothConnectionManager
+     * @param {String} address
+     * @param {Object BluetoothDevice}
+     */
+    _getPairedDeviceByAddress: 
+    function btcm__getPairedDeviceByAddress(address) {
+      if (!this._defaultAdapter) {
+        return null;
+      }
+
+      var pairedDevices = this._defaultAdapter.getPairedDevices();
+      if (pairedDevices.length === 0) {
+        return null;
+      } else {
+        for (var i in pairedDevices) {
+          if (pairedDevices[i].address === address) {
+            return pairedDevices[i];
+          }
+        }
+        return null;
+      }
+    },
+
+    /**
+     * Find out there is any profile still connected.
+     *
+     * @access private
+     * @memberOf BluetoothConnectionManager
+     * @param {String} address - the address of device
+     */
+    _hasConnectedProfileByAddress: 
+    function btcm__hasConnectedProfileByAddress(address) {
+      if (!this._connectedDevicesInfo[address]) {
+        return false;
+      }
+
+      var hasConnectedProfile = false;
+      for (var profileID in this.Profiles) {
+        var connectedProfiles = 
+          this._connectedDevicesInfo[address].connectedProfiles;
+        if (connectedProfiles && (connectedProfiles[profileID] === true)) {
+          hasConnectedProfile = true;
+        }
+      }
+      return hasConnectedProfile;
+    },
+
+    /**
+     * When device is connecting, we will call all registered listeners.
+     *
+     * @memberOf BluetoothConnectionManager
+     * @param {Event} event
+     */
+    _deviceConnecting: function btcm__deviceConnected(event) {
```

Comment:
Following the previous comment, the four functions for emitting events could be combined. Note that the object passed to the listener should contain a `type` property so the callback can handle the event more easily. We can use the structure borrowed from html events like:
```
{
  type: 'connecting',
  detail: { // the object you want to pass to the listener }
}
```